helix-core/src/movement.rs

```diff
@@ -1,18 +1,60 @@
-use crate::graphemes::{nth_next_grapheme_boundary, nth_prev_grapheme_boundary, RopeGraphemes};
-use crate::{coords_at_pos, pos_at_coords, ChangeSet, Position, Range, Rope, RopeSlice, Selection};
+use std::iter;
+
+use crate::{
+    coords_at_pos,
+    graphemes::{nth_next_grapheme_boundary, nth_prev_grapheme_boundary, RopeGraphemes},
+    pos_at_coords, ChangeSet, Position, Range, Rope, RopeSlice, Selection,
+};
 
 #[derive(Copy, Clone, PartialEq, Eq)]
 pub enum Direction {
     Forward,
     Backward,
 }
 
+#[derive(Copy, Clone, PartialEq, Eq)]
+pub enum SelectionBehaviour {
+    Extend,
+    Displace,
+}
+
+pub trait SliceHelpers {
+    fn outside(&self, text: RopeSlice) -> bool;
+    /// The next character after this belongs
+    /// to a different `Category`
+    fn is_category_boundary(&self, text: RopeSlice) -> bool;
+    fn is(&self, category: Category, text: RopeSlice) -> bool;
+    fn category(&self, text: RopeSlice) -> Option<Category>;
+}
+
+impl SliceHelpers for usize {
+    fn outside(&self, text: RopeSlice) -> bool {
```

Comment:
Could these methods be `inline`d?

Reply:
I usually don't mark private functions as `inline` since you're not giving the compiler that much to work with, since there's no decision to be made with regards to inlining across crate boundaries. If they end up becoming public and we're not using LTO though, why not.